app/plugin_api.go

```diff
@@ -929,3 +929,125 @@ func (api *PluginAPI) PluginHTTP(request *http.Request) *http.Response {
 	api.app.ServeInterPluginRequest(responseTransfer, request, api.id, destinationPluginId)
 	return responseTransfer.GenerateResponse()
 }
+
+func (api *PluginAPI) CreateCommand(cmd *model.Command) (*model.Command, error) {
+	cmd.CreatorId = ""
+	cmd.PluginId = api.id
+
+	cmd, appErr := api.app.CreateCommand(cmd)
+
+	return cmd, normalizeAppErr(appErr)
+}
+
+func (api *PluginAPI) ListAllCommands(teamID string) ([]*model.Command, error) {
+	ret := make([]*model.Command, 0)
+
+	cmds, err := api.ListCustomCommands(teamID)
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, cmds...)
+
+	cmds, err = api.ListPluginCommands(teamID)
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, cmds...)
+
+	cmds, err = api.ListBuiltInCommands()
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, cmds...)
+
+	return ret, nil
+}
+
+func (api *PluginAPI) ListCustomCommands(teamID string) ([]*model.Command, error) {
+	// Plugins are allowed to bypass the a.Config().ServiceSettings.EnableCommands setting.
+	return api.app.Srv().Store.Command().GetByTeam(teamID)
+}
+
+func (api *PluginAPI) ListPluginCommands(teamID string) ([]*model.Command, error) {
+	commands := make([]*model.Command, 0)
+	seen := make(map[string]bool)
+
+	for _, cmd := range api.app.PluginCommandsForTeam(teamID) {
+		if !seen[cmd.Trigger] {
+			seen[cmd.Trigger] = true
+			commands = append(commands, cmd)
+		}
+	}
+
+	return commands, nil
+}
+
+func (api *PluginAPI) ListBuiltInCommands() ([]*model.Command, error) {
+	commands := make([]*model.Command, 0)
+	seen := make(map[string]bool)
+
+	for _, value := range commandProviders {
+		if cmd := value.GetCommand(api.app, api.app.T); cmd != nil {
+			cpy := *cmd
+			if cpy.AutoComplete && !seen[cpy.Trigger] {
+				cpy.Sanitize()
+				seen[cpy.Trigger] = true
+				commands = append(commands, &cpy)
+			}
+		}
+	}
+
+	return commands, nil
+}
+
+func (api *PluginAPI) GetCommand(commandID string) (*model.Command, error) {
```

Comment:
I know I've waffled on the terminology here, but I'm worried about the potential for confusion between `GetCommand` and `ListCustomCommands`. It's not clear from the naming that they are peers. One idea might be to keep it as `ListCommands` (and `ListPluginCommands`, etc.)

Reply:
Yep, that makes sense. Done!